File: app/src/main/java/in/hridayan/ashell/utils/Commands.java
package in.hridayan.ashell.utils;

import android.content.Context;
import android.content.res.Resources;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import in.hridayan.ashell.items.CommandItems;
import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;
import in.hridayan.ashell.R;

public class Commands {
  private static List<CommandItems> mPackages = null;

  // Define a model for JSON parsing
  private static class CommandModel {
    String command;
    String example;
        String description;
  }

  // Load commands from JSON
  public static List<CommandItems> commandList(Context context) {
    List<CommandItems> mCommands = new ArrayList<>();
    try {
      // Load the JSON file from res/raw
      Resources resources = context.getResources();
      InputStream inputStream = resources.openRawResource(R.raw.commands);
      Reader reader = new InputStreamReader(inputStream);

      Gson gson = new Gson();
      Type commandListType = new TypeToken<List<CommandModel>>() {}.getType();
      List<CommandModel> commandModels = gson.fromJson(reader, commandListType);

      // Convert JSON models to CommandItems
      for (CommandModel model : commandModels) {
        mCommands.add(new CommandItems(model.command, model.example,model.description, context));
      }

      // Close the reader
      reader.close();
    } catch (Exception e) {
      e.printStackTrace();
    }
    return mCommands;
  }

  public static List<CommandItems> getCommand(String command, Context context) {
    List<CommandItems> mCommands = new ArrayList<>();
    for (CommandItems commands : commandList(context)) {
      if (commands.getTitle().contains(command)) {
        mCommands.add(commands);
      }
    }
    return mCommands;
  }

  public static List<CommandItems> getPackageInfo(String command, Context context) {
    loadPackageInfo(context);
    List<CommandItems> mCommands = new ArrayList<>();
    for (CommandItems packages : mPackages) {
      if (packages.getTitle().contains(command)) {
        mCommands.add(packages);
      }
    }
    return mCommands;
  }

  public static void loadPackageInfo(Context context) {
    mPackages = new ArrayList<>();

    try {
      Process mProcess = Runtime.getRuntime().exec("pm list packages");
      BufferedReader mInput = new BufferedReader(new InputStreamReader(mProcess.getInputStream()));
      String line;
      while ((line = mInput.readLine()) != null) {
        if (line.startsWith("package:")) {
          mPackages.add(new CommandItems(line.replace("package:", ""), null,null, context));
        }
      }
      mProcess.waitFor();
    } catch (Exception ignored) {
    }
  }
}

File: app/src/main/java/in/hridayan/ashell/utils/Commands.kt
package `in`.hridayan.ashell.utils

import android.content.Context
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import `in`.hridayan.ashell.R
import `in`.hridayan.ashell.items.CommandItems
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.Reader

object Commands {
    private lateinit var mPackages: MutableList<CommandItems>

    // Load commands from JSON
    fun commandList(context: Context): List<CommandItems> {
        val mCommands: MutableList<CommandItems> = ArrayList<CommandItems>()
        try {
            // Load the JSON file from res/raw
            val resources = context.resources
            val inputStream = resources.openRawResource(R.raw.commands)
            val reader: Reader = InputStreamReader(inputStream)

            val gson = Gson()
            val commandListType = object : TypeToken<List<CommandModel?>?>() {}.type
            val commandModels = gson.fromJson<List<CommandModel>>(reader, commandListType)

            // Convert JSON models to CommandItems
            for (model in commandModels) {
                mCommands.add(
                    CommandItems(
                        model.command,
                        model.example,
                        model.description,
                        context
                    )
                )
            }

            // Close the reader
            reader.close()
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return mCommands
    }

    @JvmStatic
    fun getCommand(command: String, context: Context): List<CommandItems> {
        val mCommands: MutableList<CommandItems> = ArrayList<CommandItems>()
        for (commands in commandList(context)) {
            if (commands.title.startsWith(command)) {
                mCommands.add(commands)
            }
        }
        return mCommands
    }

    @JvmStatic
    fun getPackageInfo(command: String, context: Context?): List<CommandItems> {
        loadPackageInfo(context)
        val mCommands: MutableList<CommandItems> = ArrayList<CommandItems>()
        for (packages in mPackages) {
            if (packages.title.startsWith(command)) {
                mCommands.add(packages)
            }
        }
        return mCommands
    }

    private fun loadPackageInfo(context: Context?) {
        mPackages = ArrayList<CommandItems>()

        try {
            val mProcess = Runtime.getRuntime().exec("pm list packages")
            val mInput = BufferedReader(InputStreamReader(mProcess.inputStream))
            var line: String
            while ((mInput.readLine().also { line = it }) != null) {
                if (line.startsWith("package:")) {
                    mPackages.add(CommandItems(line.replace("package:", ""), null, null, context))
                }
            }
            mProcess.waitFor()
        } catch (ignored: Exception) {
        }
    }

    // Define a model for JSON parsing
    private class CommandModel {
        var command: String? = null
        var example: String? = null
        var description: String? = null
    }
}
